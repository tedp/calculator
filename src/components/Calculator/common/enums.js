const KeyType = Object.freeze({
  numeric: 'numeric',
  dot: 'dot',
  operator: 'operator',
  clear: 'clear',
  result: 'result',
});

const DisplayState = Object.freeze({
  numeric: 'numeric',
  decimal: 'decimal',
  dot: 'dot',
  operator: 'operator',
});

const CalcState = Object.freeze({
  resulted: 'resulted',
  cleared: 'cleared',
  input: 'input',
});

export { KeyType, CalcState, DisplayState };
