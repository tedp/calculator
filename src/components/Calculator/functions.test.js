import { KeyType, DisplayState, CalcState } from './common/enums';
import calculateState from './functions';


describe('calculator', () => {
  let initialState;

  beforeEach(() => {
    const keyTypes = [
      KeyType.numeric,
      KeyType.operator,
      KeyType.numeric,
    ];

    const displayStates = [
      DisplayState.numeric,
      DisplayState.operator,
      DisplayState.numeric,
    ];

    initialState = {
      result: { display: '1+1', keyTypes, displayStates },
      calcState: CalcState.input,
      displayState: DisplayState.numeric,
    };
  });

  describe('calculateState', () => {
    it('should clear one character from the right', () => {
      const actual = calculateState(initialState, 'c', KeyType.clear);

      expect(actual).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+');
    });

    it('should concat operators after the caclulator has been resulted', () => {
      const state = calculateState(initialState, '=', KeyType.result);
      const actual = calculateState(state, '+', KeyType.operator);

      expect(actual).toMatchSnapshot();
      expect(actual.result.display).toEqual('2+');
    });

    it('should clear displayed value after the caclulator has been resulted when number input', () => {
      const state = calculateState(initialState, '=', KeyType.result);
      const actual = calculateState(state, '1', KeyType.numeric);

      expect(actual).toMatchSnapshot();
      expect(actual.result.display).toEqual('1');
    });

    it('should not concat numeric values after all input has been cleared', () => {
      let state = calculateState(initialState, 'c', KeyType.clear);
      state = calculateState(state, 'c', KeyType.clear);
      state = calculateState(state, 'c', KeyType.clear);
      const actual = calculateState(state, '1', KeyType.numeric);

      expect(actual).toMatchSnapshot();
      expect(actual.result.display).toEqual('1');
    });

    it('should concat numeric if current display is `1+1`', () => {
      const actual = calculateState(initialState, '1', KeyType.numeric);

      expect(actual).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+11');
    });

    it('should concat operator if current display is `1+1`', () => {
      const actual = calculateState(initialState, '+', KeyType.operator);

      expect(actual).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+1+');
    });

    it('should concat decimal if display is `1+1`', () => {
      const actual = calculateState(initialState, '.', KeyType.dot);

      expect(actual).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+1.');
    });

    it('should concat numeric if current display is `1+1+`', () => {
      const state = calculateState(initialState, '+', KeyType.operator);
      expect(state.displayState).toEqual(DisplayState.operator);

      const actual = calculateState(state, '1', KeyType.numeric);

      expect(state).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+1+1');
    });

    it('should not concat operator if display is `1+1+`', () => {
      const state = calculateState(initialState, '+', KeyType.operator);
      expect(state.result.display).toEqual('1+1+');
      expect(state.displayState).toEqual(DisplayState.operator);

      const actual = calculateState(state, '+', KeyType.operator);

      expect(state).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+1+');
    });

    it('should not concat decimal if display is `1+1+`', () => {
      const state = calculateState(initialState, '+', KeyType.operator);
      expect(state.result.display).toEqual('1+1+');
      expect(state.displayState).toEqual(DisplayState.operator);

      const actual = calculateState(state, 'dot', KeyType.dot);

      expect(state).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+1+');
    });

    it('should concat numeric if display is `1+1.`', () => {
      const state = calculateState(initialState, '.', KeyType.dot);
      expect(state.result.display).toEqual('1+1.');
      expect(state.displayState).toEqual(DisplayState.dot);

      const actual = calculateState(state, '1', KeyType.numeric);

      expect(state).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+1.1');
    });

    it('should not concat decimal if display is `1+1.`', () => {
      const state = calculateState(initialState, '.', KeyType.dot);
      expect(state.result.display).toEqual('1+1.');
      expect(state.displayState).toEqual(DisplayState.dot);

      const actual = calculateState(state, 'dot', KeyType.dot);

      expect(state).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+1.');
    });

    it('should not concat operator if display is `1+1.`', () => {
      const state = calculateState(initialState, '.', KeyType.dot);
      expect(state.result.display).toEqual('1+1.');
      expect(state.displayState).toEqual(DisplayState.dot);

      const actual = calculateState(state, '+', KeyType.operator);

      expect(state).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+1.');
    });

    it('should concat operator if display is `1+1.1`', () => {
      let state = calculateState(initialState, '.', KeyType.dot);
      state = calculateState(state, '1', KeyType.numeric);
      expect(state.result.display).toEqual('1+1.1');
      expect(state.displayState).toEqual(DisplayState.decimal);

      const actual = calculateState(state, '+', KeyType.operator);

      expect(state).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+1.1+');
    });

    it('should concat numeric if display is `1+1.1`', () => {
      let state = calculateState(initialState, '.', KeyType.dot);
      state = calculateState(state, '1', KeyType.numeric);
      expect(state.result.display).toEqual('1+1.1');
      expect(state.displayState).toEqual(DisplayState.decimal);

      const actual = calculateState(state, '1', KeyType.numeric);

      expect(state).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+1.11');
    });

    it('should not concat decimal if display is `1+1.1`', () => {
      let state = calculateState(initialState, '.', KeyType.dot);
      state = calculateState(state, '1', KeyType.numeric);
      expect(state.result.display).toEqual('1+1.1');
      expect(state.displayState).toEqual(DisplayState.decimal);

      const actual = calculateState(state, '.', KeyType.dot);

      expect(state).toMatchSnapshot();
      expect(actual.result.display).toEqual('1+1.1');
    });
  });

  describe('calulateResultState', () => {
    it('should calculate the result', () => {
      const actual = calculateState(initialState, '=', KeyType.result);

      expect(actual).toMatchSnapshot();
      expect(actual.result.display).toEqual(2);
      expect(actual.result.keyTypes).toEqual([KeyType.numeric])
      expect(actual.result.displayStates).toEqual([DisplayState.numeric])
    });

    it('should calculate the result applying the operators in the correct order `/ * - +`', () => {
      let state = calculateState(initialState, 'รท', KeyType.operator);
      state = calculateState(state, '1', KeyType.numeric);
      state = calculateState(state, 'ร', KeyType.operator);
      state = calculateState(state, '2', KeyType.numeric);
      state = calculateState(state, '-', KeyType.operator);
      state = calculateState(state, '2', KeyType.numeric);
      state = calculateState(state, '.', KeyType.dot);
      state = calculateState(state, '1', KeyType.numeric);
      const actual = calculateState(state, '=', KeyType.result);

      expect(actual).toMatchSnapshot();
      expect(actual.result.display).toEqual(0.8999999999999999);
    });
  });
});
