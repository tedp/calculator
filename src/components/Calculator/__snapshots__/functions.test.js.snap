// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator calculateState should clear displayed value after the caclulator has been resulted when number input 1`] = `
Object {
  "calcState": "input",
  "displayState": "numeric",
  "result": Object {
    "display": "1",
    "displayStates": Array [
      "numeric",
    ],
    "keyTypes": Array [
      "numeric",
    ],
  },
}
`;

exports[`calculator calculateState should clear one character from the right 1`] = `
Object {
  "calcState": "input",
  "displayState": "operator",
  "result": Object {
    "display": "1+",
    "displayStates": Array [
      "numeric",
      "operator",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
    ],
  },
}
`;

exports[`calculator calculateState should concat decimal if display is \`1+1\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "dot",
  "result": Object {
    "display": "1+1.",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
    ],
  },
}
`;

exports[`calculator calculateState should concat numeric if current display is \`1+1+\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "operator",
  "result": Object {
    "display": "1+1+1",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "operator",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "operator",
    ],
  },
}
`;

exports[`calculator calculateState should concat numeric if current display is \`1+1\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "numeric",
  "result": Object {
    "display": "1+11",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "numeric",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "numeric",
    ],
  },
}
`;

exports[`calculator calculateState should concat numeric if display is \`1+1.\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "dot",
  "result": Object {
    "display": "1+1.1",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
    ],
  },
}
`;

exports[`calculator calculateState should concat numeric if display is \`1+1.1\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "decimal",
  "result": Object {
    "display": "1+1.11",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
      "decimal",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
      "numeric",
    ],
  },
}
`;

exports[`calculator calculateState should concat operator if current display is \`1+1\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "operator",
  "result": Object {
    "display": "1+1+",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "operator",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "operator",
    ],
  },
}
`;

exports[`calculator calculateState should concat operator if display is \`1+1.1\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "decimal",
  "result": Object {
    "display": "1+1.1+",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
      "decimal",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
      "numeric",
    ],
  },
}
`;

exports[`calculator calculateState should concat operators after the caclulator has been resulted 1`] = `
Object {
  "calcState": "input",
  "displayState": "operator",
  "result": Object {
    "display": "2+",
    "displayStates": Array [
      "numeric",
      "operator",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
    ],
  },
}
`;

exports[`calculator calculateState should not concat decimal if display is \`1+1+\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "operator",
  "result": Object {
    "display": "1+1+",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "operator",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "operator",
    ],
  },
}
`;

exports[`calculator calculateState should not concat decimal if display is \`1+1.\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "dot",
  "result": Object {
    "display": "1+1.",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
    ],
  },
}
`;

exports[`calculator calculateState should not concat decimal if display is \`1+1.1\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "decimal",
  "result": Object {
    "display": "1+1.1",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
      "decimal",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
      "numeric",
    ],
  },
}
`;

exports[`calculator calculateState should not concat numeric values after all input has been cleared 1`] = `
Object {
  "calcState": "input",
  "displayState": "numeric",
  "result": Object {
    "display": "1",
    "displayStates": Array [
      "numeric",
    ],
    "keyTypes": Array [
      "numeric",
    ],
  },
}
`;

exports[`calculator calculateState should not concat operator if display is \`1+1+\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "operator",
  "result": Object {
    "display": "1+1+",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "operator",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "operator",
    ],
  },
}
`;

exports[`calculator calculateState should not concat operator if display is \`1+1.\` 1`] = `
Object {
  "calcState": "input",
  "displayState": "dot",
  "result": Object {
    "display": "1+1.",
    "displayStates": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
    ],
    "keyTypes": Array [
      "numeric",
      "operator",
      "numeric",
      "dot",
    ],
  },
}
`;

exports[`calculator calulateResultState should calculate the result 1`] = `
Object {
  "calcState": "resulted",
  "displayState": "numeric",
  "result": Object {
    "display": 2,
    "displayStates": Array [
      "numeric",
    ],
    "keyTypes": Array [
      "numeric",
    ],
  },
}
`;

exports[`calculator calulateResultState should calculate the result applying the operators in the correct order \`/ * - +\` 1`] = `
Object {
  "calcState": "resulted",
  "displayState": "numeric",
  "result": Object {
    "display": 0.8999999999999999,
    "displayStates": Array [
      "numeric",
    ],
    "keyTypes": Array [
      "numeric",
    ],
  },
}
`;
